src/Servers/Kestrel/Core/src/Internal/Http2/Http2KeepAlive.cs

```diff
@@ -0,0 +1,116 @@
+        PingSent,
+        Timeout
+    }
+
+    internal class Http2KeepAlive
+    {
+        internal static readonly ReadOnlySequence<byte> PingPayload = new ReadOnlySequence<byte>(new byte[8]);
+
+        private readonly TimeSpan _keepAliveInterval;
+        private readonly TimeSpan _keepAliveTimeout;
+        private readonly ISystemClock _systemClock;
+        private readonly object _lock = new object();
+        private KeepAliveState _state;
+        private bool _bytesReceivedCurrentTick;
+        private long _lastBytesReceivedTimestamp;
+        private long _pingSentTimestamp;
+
+        public Http2KeepAlive(TimeSpan keepAliveInterval, TimeSpan keepAliveTimeout, ISystemClock systemClock)
+        {
+            _keepAliveInterval = keepAliveInterval;
+            _keepAliveTimeout = keepAliveTimeout;
+            _systemClock = systemClock;
+        }
+
+        internal KeepAliveState ProcessKeepAlive(bool dataReceived)
+        {
+            lock (_lock)
+            {
+                if (dataReceived)
+                {
+                    _bytesReceivedCurrentTick = true;
+                }
+                return _state;
+            }
+        }
+
+        public void PingSent()
+        {
+            lock (_lock)
+            {
+                _state = KeepAliveState.PingSent;
+                _pingSentTimestamp = _systemClock.UtcNowTicks;
+            }
+        }
+
+        public void PingAckReceived()
+        {
+            lock (_lock)
+            {
+                if (_state == KeepAliveState.PingSent)
+                {
+                    _pingSentTimestamp = 0;
+                    _state = KeepAliveState.None;
+                }
+            }
+        }
+
+        public void Tick(DateTimeOffset now)
+        {
+            var timestamp = now.Ticks;
+
+            lock (_lock)
+            {
+                // Bytes were received since the last tick.
+                // Update a timestamp of when bytes were last received.
+                if (_bytesReceivedCurrentTick)
+                {
+                    _lastBytesReceivedTimestamp = timestamp;
+                    _bytesReceivedCurrentTick = false;
+                }
+
+                switch (_state)
+                {
+                    case KeepAliveState.None:
+                        // Check whether keep alive interval has passed since last bytes received
+                        if (timestamp > (_lastBytesReceivedTimestamp + _keepAliveInterval.Ticks))
+                        {
+                            _state = KeepAliveState.SendPing;
+                        }
+                        return;
+                    case KeepAliveState.SendPing:
+                        return;
+                    case KeepAliveState.PingSent:
+                        if (_keepAliveTimeout != Timeout.InfiniteTimeSpan)
+                        {
+                            if (timestamp > (_pingSentTimestamp + _keepAliveTimeout.Ticks))
```

Comment:
Nit: [Kestrel's ISystemClock](https://github.com/dotnet/aspnetcore/blob/52fdd8f2a2f22ac84ecfb268f6bda6d75976c230/src/Servers/Kestrel/Core/src/Internal/Infrastructure/HeartbeatManager.cs) only has a second of precision. This means that `_pingSentTimestamp` can be up to a second earlier than it should be. I'd rather error on timing out too late than too soon, so we should add a a second to the expiration timestamp.

Double nit: We can compute the expiration timestamp once instead of per tick.